lib/src/main/scala/code/chymyst/jc/ReactionSite.scala

```diff
@@ -287,50 +294,60 @@ private[jc] final class ReactionSite(reactions: Seq[Reaction], reactionPool: Poo
+    replyValueWrapper.resetSemaphoreForReply()
+    replyValueWrapper.releaseSemaphoreForReply()
     val blockingMolValue = BlockingMolValue(v, replyValueWrapper)
     emit(bm, blockingMolValue)
     val success =
       BlockingIdle {
         replyValueWrapper.acquireSemaphore(timeoutNanos = timeoutOpt)
       }
-    replyValueWrapper.deleteSemaphore()
     // We might have timed out, in which case we need to forcibly remove the blocking molecule from the soup.
     removeBlockingMolecule(bm, blockingMolValue, !success)
 
-    success
+    val result = replyValueWrapper.replyStatus
+    replyValueWrapper.releaseSemaphoreForReply()
+
+    result
   }
 
   // Adding a blocking molecule may trigger at most one reaction and must return a value of type R.
   // We must make this a blocking call, so we acquire a semaphore (with or without timeout).
   private[jc] def emitAndAwaitReply[T,R](bm: BlockingMolecule[T,R], v: T, replyValueWrapper: AbsReplyValue[T,R]): R = {
-    emitAndAwaitReplyInternal(timeoutOpt = None, bm, v, replyValueWrapper)
+    val result = emitAndAwaitReplyInternal(timeoutOpt = None, bm, v, replyValueWrapper)
     // check if we had any errors, and that we have a result value
-    replyValueWrapper.errorMessage match {
-      case Some(message) => throw new Exception(message)
-      case None => replyValueWrapper.result.getOrElse(
-        throw new ExceptionEmptyReply(s"Internal error: In $this: $bm received an empty reply without an error message"
-        )
-      )
+    result match {
+      case ErrorNoReply(message) => throw new Exception(message)
+      case HaveReply(res) => res.asInstanceOf[R]
+      case _ => throw new Exception(s"In $this: Internal error: reply status for $bm is $result")
     }
   }
 
-  // This is a separate method because it has a different return type than emitAndReply.
-  private[jc] def emitAndAwaitReplyWithTimeout[T,R](timeout: Long, m: BlockingMolecule[T,R], v: T, replyValueWrapper: AbsReplyValue[T,R]):
+  // This is a separate method because it has a different return type than [[emitAndAwaitReply]].
+  private[jc] def emitAndAwaitReplyWithTimeout[T,R](timeout: Long, bm: BlockingMolecule[T,R], v: T, replyValueWrapper: AbsReplyValue[T,R]):
   Option[R] = {
-    val haveReply = emitAndAwaitReplyInternal(timeoutOpt = Some(timeout), m, v, replyValueWrapper)
+    val result = emitAndAwaitReplyInternal(timeoutOpt = Some(timeout), bm, v, replyValueWrapper)
     // check if we had any errors, and that we have a result value
-    replyValueWrapper.errorMessage match {
-      case Some(message) => throw new Exception(message)
-      case None => if (haveReply) Some(replyValueWrapper.result.getOrElse(
-        throw new ExceptionEmptyReply(s"Internal error: In $this: $m received an empty reply without an error message"))
-      )
-      else None
+    result match {
+      case ErrorNoReply(message) => throw new Exception(message)
+      case HaveReply(res) => Some(res.asInstanceOf[R])
```

Comment:
That won't work, because this syntax means something else: 
```
case HaveReply(`res`) => ...
```
will match only if `HaveReply(x)` has `x == res`.